pkg/meta/distributed_lock.go

```diff
@@ -0,0 +1,314 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package meta
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/pkg/parser/model"
+	"github.com/pingcap/tidb/pkg/util/intest"
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	"go.uber.org/zap"
+)
+
+// DataStore is an interface to create transaction.
+type DataStore interface {
+	Begin(ctx context.Context) (txn DataTxn, err error)
+}
+
+// DataTxn is an interface to operate data in transaction.
+type DataTxn interface {
+	Rollback()
+	Commit() error
+
+	StartTS() uint64
+	Set(key, value []byte) error
+	Get(key []byte) ([]byte, error)
+	Del(key []byte) error
+}
+
+type lockContent struct {
+	StartTS    uint64 `json:"start_ts"`
+	InstanceID string `json:"instance_id"`
+}
+
+func newLockContent(startTS uint64, instanceID string) lockContent {
+	return lockContent{
+		StartTS:    startTS,
+		InstanceID: instanceID,
+	}
+}
+
+func (c lockContent) encode() ([]byte, error) {
+	return json.Marshal(c)
+}
+
+func (c *lockContent) decode(data []byte) error {
+	return json.Unmarshal(data, c)
+}
+
+// DistributedLock is a distributed lock implementation based on meta key transaction.
+type DistributedLock struct {
+	outerCtx   context.Context
+	store      DataStore
+	instanceID string
+
+	lockName []byte
+	backoff  time.Duration
+	lease    time.Duration
+
+	cancelRenewLoop context.CancelFunc
+}
+
+// DistributedLockBuilder is used to build a DistributedLock.
+type DistributedLockBuilder struct {
+	backoff time.Duration
+	lease   time.Duration
+}
+
+// NewDistributedLockBuilder creates a new DistributedLockBuilder.
+func NewDistributedLockBuilder() *DistributedLockBuilder {
+	return &DistributedLockBuilder{
+		backoff: 3 * time.Second,
+		lease:   30 * time.Second,
+	}
+}
+
+// SetBackoff sets the backoff time between two lock operations.
+// Default is 3 * time.Second.
+func (b *DistributedLockBuilder) SetBackoff(backoff time.Duration) *DistributedLockBuilder {
+	b.backoff = backoff
+	return b
+}
+
+// SetLease sets the lease time of the lock.
+// Default is 30 * time.Second.
+func (b *DistributedLockBuilder) SetLease(lease time.Duration) *DistributedLockBuilder {
+	b.lease = lease
+	return b
+}
+
+// Build builds a DistributedLock.
+func (b *DistributedLockBuilder) Build(
+	ctx context.Context, store DataStore, instanceID string, lockName string) *DistributedLock {
+	ctx = logutil.WithCategory(ctx, "distributed-lock")
+	return &DistributedLock{
+		outerCtx:   ctx,
+		store:      store,
+		instanceID: instanceID,
+
+		lockName: []byte(lockName),
+		backoff:  b.backoff,
+		lease:    b.lease,
+	}
+}
+
+func (d *DistributedLock) renewLockLoop(ctx context.Context) {
+	ticker := time.NewTicker(d.lease / 3)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			err := d.renew(ctx)
```

Comment:
In order to survive the temporary network error, I think we should only exit when the error is "lock is unexpectedly removed" case.

Reply:
`d.renew` itself also has retry logic.